### PR TITLE
refactor: expose ModuleDatabaseTransaction for modules

### PR DESCRIPTION
**Pull request by m1sterc001guy**
Today we wrap `IsolatedDatabaseTransaction` inside `DatabaseTransaction` and only expose `DatabaseTransaction` to the modules. This has some downsides, such as the fact that `commit_tx` is exposed as a possible function for the modules to call. This should not be allowed though, since modules should not be allowed to commit to the database because the fedimint server is responsible for managing the lifetime of a database transaction. Today if this happens, `commit_tx` will panic.

This PR is a refactor to expose `ModuleDatabaseTransaction` to the modules and add the public facing database transaction API to `ModuleDatabaseTransaction`, but removed the functions that modules should not be allowed to call. This allows us to keep the wrapper functionality that we have introduced in previous PRs, but also address the issue of not exposing the full database transaction API.

Fixes: https://github.com/fedimint/fedimint/issues/1729

### CODE REVIEW

File: fedimint-core/src/core/server.rs [hunk 1]
```diff
@@ -204,14 +211,17 @@ where
     }
 
     /// Blocks until a new `consensus_proposal` is available.
-    async fn await_consensus_proposal(&self, dbtx: &mut DatabaseTransaction<'_>) {
```
Comment (dpc) [inline fedimint-core/src/core/server.rs:207]: Do we want IsolatedDatabaseTransaction here? It is exposing outside the implementation detail of modules getting isolated. Is it an attempt to make sure we never forget about isolation on the caller side?

Comment (dpc) [reply fedimint-core/src/core/server.rs:207]: Oh. I thought I'm looking at a different PR of yours ...

**Comment by dpc** [comment]
Seems to me that a proper way to handle this is to have have separate newtype that just doesn't expose `commit_tx`. 

This way user (module developer) code is not tied to a concrete type and an implementation detail.

**Comment by m1sterc001guy** [comment]
> Seems to me that a proper way to handle this is to have have separate newtype that just doesn't expose `commit_tx`.
> 
> This way user (module developer) code is not tied to a concrete type and an implementation detail.

But the implementation details of `IsolatedDatabaseTransaction` are hidden right? I believe the only public functions are `get_value`, `insert_entry`, `insert_new_entry`, `remove_entry`, `find_by_prefix`, and `remove_by_prefix`. I don't think the modules can call the raw functions, which is where the details on how the prefix is appended/prepended are handled.

**Comment by dpc** [comment]
> But the implementation details of `IsolatedDatabaseTransaction` are hidden right?

Yes, but the `IsolatedDatabaseTransaction` has the implementation details kind of right there in the name. :)

**Comment by m1sterc001guy** [comment]
> Yes, but the IsolatedDatabaseTransaction has the implementation details kind of right there in the name. :)

Do you think just renaming `IsolatedDatabaseTransaction` would be sufficient? I suppose we could introduce a new type but I'm worried too many wrapper types will become confusing.

Perhaps:

`IsolatedDatabaseTransaction` becomes `ModuleDatabaseTransaction`
The existing `ModuleDatabaseTransaction` becomes `WrappedIsolatedDatabaseTransaction`. This will be private since it is only used inside of fedimint_core::db::mod.rs.

I'm terrible at coming up with names so those are probably still not the best.

**Comment by dpc** [comment]
> `IsolatedDatabaseTransaction` becomes `ModuleDatabaseTransaction`
> The existing `ModuleDatabaseTransaction` becomes `WrappedIsolatedDatabaseTransaction`. This will be private since it is only used inside of fedimint_core::db::mod.rs.

`ModuleDatabaseTransaction` sounds reasonable to me. Though my main concern is just avoiding tying the type we use in the module API to any concrete implementation, as in the future we might want to add some other functionality etc. that would be otherwise invisible to the user.

**Comment by elsirion** [comment]
Needs rebase after #1810

**Comment by m1sterc001guy** [comment]
> Though my main concern is just avoiding tying the type we use in the module API to any concrete implementation, as in the future we might want to add some other functionality etc. that would be otherwise invisible to the user.

Ok I'm decently happy with the most recent renames. Here is a description of each type and its purpose:

`DatabaseTransaction`: existing public type, used for atomic db transactions across modules
`ModuleDatabaseTransaction`: new public type, cannot commit to db, doesn't contain implementation details on isolation
`IsolatedDatabaseTransaction`: existing private type, contains implementation details on prepending/striping prefixes
`CommitableIsolatedDatabaseTransaction`: renamed private type, used for atomic db transactions within a module. This type will always be wrapped inside `DatabaseTransaction` so modules aren't aware of it.

Comments in code have been updated

**Review by dpc — APPROVED** [review]

File: fedimint-client/src/sm/state.rs [hunk 2]
```diff
@@ -84,7 +84,7 @@ type TriggerFuture = Pin<Box<dyn Future<Output = serde_json::Value> + Send + 'st
 // TODO: remove Arc, maybe make it a fn pointer?
 type StateTransitionFunction<S> = Arc<
     dyn for<'a> Fn(
-            &'a mut IsolatedDatabaseTransaction<'_, '_, ModuleInstanceId>,
+            &'a mut ModuleDatabaseTransaction<'_, ModuleInstanceId>,
```
Comment (elsirion) [inline fedimint-client/src/sm/state.rs:87]: nit: `ModuleDatabaseTrabsaction` could be a type def instead:

```rust
type ModuleDatabaseTransaction<'a, 'b> = IsolatedDatabaseTransaction<'a, 'b, ModuleInstanceId>;
```

**Review by elsirion — APPROVED** [review]

File: fedimint-client/src/sm/state.rs (hunk 2 above)
Comment (dpc) [reply fedimint-client/src/sm/state.rs:87]: That might be a little bit leaky though.

**Pull request merged**